Adding a new e2e test to verify volume provisioning throws error if non-shared DS is used

shaominchen opened the pull request:
**What this PR does / why we need it**:
Adding a new e2e test to verify volume provisioning throws error if non-shared DS is used

**Which issue(s) this PR fixes**:
Fixes #372 

**Testing Done**:

**1. When environment variable LOCAL_DATASTORE is not set**

$ go run hack/e2e.go --check-version-skew=false -v --test --test_args='--ginkgo.focus=Volume\sProvisioning\son\sDatastore'
......
[sig-storage] Volume Provisioning on Datastore [Feature:vsphere] 
  should fail provisioning pv on local datastore
  
[It] should fail provisioning pv on local datastore
STEP: Invoking Test for local datastore
Jan 19 14:47:41.322: INFO: Environment variable VSPHERE_LOCAL_DATASTORE is not set. Skipping the test.
S [SKIPPING] [6.832 seconds]
......
Ran 1 of 717 Specs in 135.753 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 716 Skipped PASS

**2. When environment variable LOCAL_DATASTORE is set**

$ go run hack/e2e.go --check-version-skew=false -v --test --test_args='--ginkgo.focus=Volume\sProvisioning\son\sDatastore'
......
[It] should fail provisioning pv on local datastore
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/vsphere/vsphere_volume_datastore.go:81
STEP: Invoking Test for local datastore
STEP: Creating Storage Class with given datastore
STEP: Creating PVC using the Storage Class
STEP: Expect claim to fail provisioning volume
......
Ran 2 of 717 Specs in 255.194 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 715 Skipped PASS

divyenpatel commented on test/e2e/storage/vsphere/vsphere_volume_datastore.go:
```diff
@@ -32,11 +32,13 @@ import (
 
 const (
 	InvalidDatastore = "invalidDatastore"
+	LocalDatastore   = "localDatastore" // Pre-exists in the testbed
```
Instead of defining as const, can we read non-shared local datastore name from environment variable
```
os.Getenv("LOCALDATASTORE") 
```

shaominchen replied:
Make sense. Will fix it.

prashima replied:
Is there are scope for adding this property in our existing conf file? I would prefer us to not to have another environment variable.

shaominchen replied:
As discussed with @divyenpatel offline, right now we don't plan to add any test specific environment variables into vsphere.conf file. E2e test will copy and use vsphere.conf file as is without any change.

shaominchen commented on test/e2e/storage/vsphere/vsphere_volume_datastore.go:
```diff
@@ -32,11 +32,13 @@ import (
 
 /*
-	Test to verify datastore specified in storage-class is being honored while volume creation.
+	Test case 1: Verify dynamic volume creation with an invalid datastore specified in storage class
```
Fixed.

prashima commented on test/e2e/storage/vsphere/vsphere_volume_datastore.go:
```diff
@@ -67,17 +69,29 @@ var _ = utils.SIGDescribe("Volume Provisioning on Datastore [Feature:vsphere]",
 		}
 	})
+
+	It("verify dynamically provisioned pv using storageclass fails on a local datastore", func() {
```
Can we have smaller It message? Writing regex for this It would become simpler.

shaominchen replied:
Will change it to: "should fail provisioning pv on local datastore".

shaominchen commented:
@divyenpatel I've addressed your comments. Please note that this test will be skipped since environment variable LOCAL_DATASTORE is not set as of now.

prashima commented on test/e2e/storage/vsphere/vsphere_volume_datastore.go:
```diff
@@ -67,17 +71,33 @@ var _ = utils.SIGDescribe("Volume Provisioning on Datastore [Feature:vsphere]",
+		if localDatastore == "" {
+			framework.Skipf("Environment variable LOCAL_DATASTORE is not set. Skipping the test.")
+		}
+		scParameters[Datastore] = localDatastore
```
I would recommend printing value of localDatastore variable. A wrongly set environment variable would go unnoticed otherwise.

shaominchen replied:
Add a log message for this.

prashima commented on test/e2e/storage/vsphere/vsphere_volume_datastore.go:
```diff
@@ -67,17 +71,33 @@ var _ = utils.SIGDescribe("Volume Provisioning on Datastore [Feature:vsphere]",
+		scParameters[Datastore] = localDatastore
+		scParameters[DiskFormat] = ThinDisk
+		err := invokeDatastoreTestNeg(client, namespace, scParameters)
+		Expect(err).To(HaveOccurred())
```
Is there a way to check for specific error? How do we differentiate error caused by non-shared datastore from errors due to any bug in the production code or misspelled datastore name?

shaominchen replied:
I didn't find a way to check for specific error. Different errors seems to cause the same "Status" and "Reason":
    Status: Pending
    Reason: ProvisioningFailed
Only "Message" field can tell us more detailed information.

shaominchen commented:
@divyenpatel @prashima Your comments have been addressed, please take a look.

divyenpatel commented on test/e2e/storage/vsphere/vsphere_volume_datastore.go:
```diff
@@ -63,21 +67,39 @@ var _ = utils.SIGDescribe("Volume Provisioning on Datastore [Feature:vsphere]",
+
+	It("should fail provisioning pv on local datastore", func() {
+		By("Invoking Test for local datastore")
+		localDatastore = os.Getenv("LOCAL_DATASTORE")
```
Can we use `GetAndExpectStringEnvVar` func here? 
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/vsphere/vsphere_common.go#L55

We should also move localDatastore to 
https://github.com/vmware/kubernetes/blob/master/test/e2e/storage/vsphere/vsphere_common.go#L25

shaominchen replied:
Above vsphere_common code was just merged 16 hours ago - it is not available in the current vmware/multi-vc-e2e branch. I will sync up the multi-vc-e2e branch with kubernetes/master to get those changes.

prashima replied:
I am wondering why do we need GetAndExpect here. What practice do we follow for our tests in case some env variable is not set- do we skip the test or do we fail it?

divyenpatel replied:
@prashima we can fail or skip test depending on where the check is made. 

If check is made in `BeforeEach` and all `It` specs need this env variable to be set, we can skip running all tests, else we can also check this on individual bases in the 'It' spec and fail specific test.

shaominchen replied:
I don't think we should fail or skip the test depending on where the check is made. The test should always fail for production issue and skip for misconfiguration. I see a lot of existing tests fail for env variable not set, which is incorrect and we should fix that.

divyenpatel replied:
@shaominchen suggested we should use `framework.skip` if environment is not set.
We may need to change - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/vsphere/vsphere_common.go#L64

prashima commented:
LGTM
